File: xaiop-sdk/python/src/xaiop/fragment.py
"""Root fragment representation."""

from __future__ import annotations

import json
from typing import Any


class XaiopFragment:
    """Named bindings at Root without an anonymous outer object."""

    __slots__ = ("entries",)

    def __init__(self, entries: dict[str, Any]) -> None:
        self.entries = entries

    @property
    def is_fragment(self) -> bool:
        return True

    def notation(self) -> str:
        return ",".join(
            f"{json.dumps(k)}:{json.dumps(v)}" for k, v in self.entries.items()
        )

File: xaiop-sdk/go/xaiop/fragment.go
package xaiop

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"
)

// Fragment holds named bindings at root without an anonymous outer object.
type Fragment struct {
	Entries map[string]any
}

// IsFragment reports that this value is a fragment root.
func (f *Fragment) IsFragment() bool {
	return true
}

// Notation returns a compact JSON-like summary of fragment entries.
func (f *Fragment) Notation() string {
	if f == nil || f.Entries == nil {
		return ""
	}
	keys := make([]string, 0, len(f.Entries))
	for k := range f.Entries {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	parts := make([]string, 0, len(keys))
	for _, k := range keys {
		v := f.Entries[k]
		kj, _ := json.Marshal(k)
		vj, _ := json.Marshal(v)
		parts = append(parts, fmt.Sprintf("%s:%s", string(kj), string(vj)))
	}
	return strings.Join(parts, ",")
}
